crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -457,36 +326,270 @@ impl Drop for AbortOnDrop {
+
+                match response.status {
+                    ResponseStatus::Ack => Ok(response.sequence_number),
+                    status => Err(anyhow::anyhow!(
+                        "failed to query latest dedup \
+                        sequence number for '{legacy_producer_id}': {status:?}"
+                    )),
+                }
+            },
+            |err| {
+                debug!("Failed to query legacy dedup information: {err} .. retrying");
+            },
+        )
+        .await
+        .expect("tries forever")
+    }
+
+    async fn run_inner(&mut self) -> Result<(), Error> {
+        debug!(
+            restate.subscription.id = %self.builder.subscription().id(),
+            messaging.consumer.group.name = self.consumer_group_id,
+            "Starting topic '{}' partition '{}' consumption loop",
+            self.topic_partition.0,
+            self.topic_partition.1
+        );
+
+        let legacy_dedup_offset = self.legacy_dedup_offset().await;
+        debug!(
+            topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1,
+            consumer_group=%self.consumer_group_id,
+            "Legacy dedup offset: {legacy_dedup_offset:?}",
+        );
+
+        let producer_id = dedup_producer_id(
+            &self.builder.subscription().id(),
+            &self.consumer_group_id,
+            &self.topic_partition.0,
+            self.topic_partition.1,
+        );
+
+        let ingress_request_counter = counter!(
+            KAFKA_INGRESS_REQUESTS,
+            "subscription" => self.builder.subscription().id().to_string(),
+            "topic" => self.topic_partition.0.to_string(),
+            "partition" => self.topic_partition.1.to_string(),
+        );
+
+        let mut inflight = VecDeque::new();
+
+        let mut consumer_stream = self.topic_partition_consumer.stream();
+        loop {
+            tokio::select! {
+                biased;
+                Some(committed) = Self::head_committed(&mut inflight) => {
+                    _ = inflight.pop_front().expect("to exist");
+                    let offset = committed.map_err(|_| Error::IngressClosed)?;
+
+                    ingress_request_counter.increment(1);
+                    trace!(
+                        topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1,
+                        offset=%offset, consumer_group=%self.consumer_group_id,
+                        "store kafka offset",
+                    );
+                    self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, offset)?;
+                },
+                Some(received) = consumer_stream.next() => {
+                    let msg = received?;
+                    let offset = msg.offset();
+                    if let Some(dedup_offset) = legacy_dedup_offset && offset as u64 <= dedup_offset {
+                        // we skip the messages instead of seeking to avoid seeking backward,
```

Comment:
What do you mean with seeking backward? Do you mean the case, where the consumer has already read beyond the `dedup_offset` point and seeking to `dedup_offset` would mean seeking backward?